lab_2/main.py

```diff
@@ -132,13 +304,55 @@ def find_lcs_length_optimized(first_sentence_tokens: tuple, second_sentence_toke
     :param plagiarism_threshold: a threshold
     :return: a length of the longest common subsequence
     """
-    pass
+    if not isinstance(first_sentence_tokens, tuple) or not isinstance(second_sentence_tokens, tuple) or \
+            not isinstance(plagiarism_threshold, float):
+        return -1
+    if None in first_sentence_tokens or None in second_sentence_tokens or \
+            plagiarism_threshold < 0 or plagiarism_threshold > 1:
+        return -1
+    if len(first_sentence_tokens) == 0 or len(second_sentence_tokens) == 0:
+        return 0
+    if len(first_sentence_tokens) > len(second_sentence_tokens):
+        len_search = len(second_sentence_tokens)
+    elif len(second_sentence_tokens) > len(first_sentence_tokens):
+        len_search = len(first_sentence_tokens)
+    else:
+        len_search = len(second_sentence_tokens)
+    current_row = [0] * len_search
+    for row, word_1 in enumerate(first_sentence_tokens[:len_search]):
+        previous_row = current_row[:]
+        for column, word_2 in enumerate(second_sentence_tokens[:len_search]):
+            if row == column:
+                if word_1 == word_2:
+                    current_row[column] = previous_row[column - 1] + 1
+            else:
+                current_row[column] = max((current_row[column - 1], previous_row[column]))
+    lcs_length = current_row[-1]
+    if lcs_length / len(second_sentence_tokens) < plagiarism_threshold:
+        return 0
+    return lcs_length
 
 
-def tokenize_big_file(path_to_file: str) -> tuple:
+def tokenize_big_file(path_to_file: str, id_number=1) -> tuple:
     """
     Reads, tokenizes and transforms a big file into a numeric form
     :param path_to_file: a path
+    :param id_number: an id
     :return: a tuple with ids
     """
-    pass
+    tokens = []
+    id_dict = {}
+
+    def get_id(word):
+        nonlocal id_number
+        if word not in id_dict:
+            id_dict[word] = id_number
+            id_number += 1
+        return id_dict[word]
+    with open(path_to_file, encoding='UTF-8') as file:
+        for line in file.readlines():
```

Comment:
still, a file is read at once
you need to, for example, read first 1000 lines, tokenize them, transform into a numeric form and only then read other 1000 lines